libs/ui/src/hooks/smartcard_auth/use_inserted_smartcard_auth.ts

```diff
@@ -379,3 +383,117 @@ export function useInsertedSmartcardAuth({
       throwIllegalValue(auth, 'status');
   }
 }
+
+async function logAuthEvents(
+  logger: Logger,
+  previousAuth: InsertedSmartcardAuth.Auth = {
+    status: 'logged_out',
+    reason: 'no_card',
+  },
+  auth: InsertedSmartcardAuth.Auth
+) {
+  switch (previousAuth.status) {
+    case 'logged_out': {
+      if (auth.status === 'logged_in') {
+        // TODO do we want to log any user attributes?
+        // e.g. for cardless voters, precinctId and ballotStyleId
+        await logger.log(LogEventId.AuthLogin, auth.user.role, {
+          disposition: LogDispositionStandardTypes.Success,
+        });
+      } else if (
+        previousAuth.reason === 'no_card' &&
+        auth.status === 'logged_out' &&
+        auth.reason !== 'no_card'
+      ) {
+        await logger.log(LogEventId.AuthLogin, auth.cardUserRole ?? 'unknown', {
+          disposition: LogDispositionStandardTypes.Failure,
+          message: `User failed login: ${auth.reason}`,
+          reason: auth.reason,
+        });
+      }
+      return;
+    }
+
+    case 'checking_passcode': {
+      if (auth.status === 'logged_out') {
+        await logger.log(LogEventId.AuthPasscodeEntry, previousAuth.user.role, {
+          disposition: LogDispositionStandardTypes.Failure,
+          message: 'User canceled passcode entry.',
+        });
+      } else if (auth.status === 'logged_in') {
+        await logger.log(LogEventId.AuthPasscodeEntry, auth.user.role, {
+          disposition: LogDispositionStandardTypes.Success,
+          message: 'User entered correct passcode.',
+        });
+        await logger.log(LogEventId.AuthLogin, auth.user.role, {
+          disposition: LogDispositionStandardTypes.Success,
+        });
+      } else if (auth.status === 'checking_passcode') {
+        if (
+          auth.wrongPasscodeEnteredAt &&
+          previousAuth.wrongPasscodeEnteredAt !== auth.wrongPasscodeEnteredAt
+        ) {
+          await logger.log(
+            LogEventId.AuthPasscodeEntry,
+            previousAuth.user.role,
+            {
+              disposition: LogDispositionStandardTypes.Failure,
+              message: 'User entered incorrect passcode.',
+            }
+          );
+        }
+      }
+      return;
+    }
+
+    case 'logged_in': {
+      if (auth.status === 'logged_out') {
+        await logger.log(LogEventId.AuthLogout, previousAuth.user.role, {
+          disposition:
+            auth.reason === 'no_card'
```

Comment:
what are the reasons where this ends in a failure?

Reply:
Generally, cases where the user is logged in and then something happens that forces them to get logged out. I think currently the only case is when a voter card expires.

Reply:
Do you think that `failure` is the right disposition for that case? Still wrapping my head around things